compiler/plugins/input/Torch/InputConversion/ConvertTMTensorToLinalgExt.cpp

```diff
@@ -141,16 +153,35 @@ struct AttentionOpConversion
     }
 
     // Attention only works for FloatType.
+    auto Q_type = op.getQueryType().getElementType();
+    auto bit_width = Q_type.getIntOrFloatBitWidth();
+    Value scaled_query;
+    double dk = static_cast<double>(headDim);
     FloatType targetType = cast<FloatType>(op.getQueryType().getElementType());
 
-    double dk = static_cast<double>(headDim);
-    dk = 1.0 / std::sqrt(dk);
-    Value scale = rewriter.create<arith::ConstantOp>(
-        loc, targetType, rewriter.getFloatAttr(targetType, dk));
+    // Since we use exp2 for attention instead of the original exp, we have to
+    // multiply the scale by log2(e). We use exp2 instead of exp as most
+    // platforms have better support for exp2 (we verified that we gain some
+    // speedup on some GPUs).
+    if (bit_width >= 16) {
+      dk = (1.0 / std::sqrt(dk)) * M_LOG2E;
+      Value scale = rewriter.create<arith::ConstantOp>(
+          loc, targetType, rewriter.getFloatAttr(targetType, dk));
+      Value scaleTensor =
+          rewriter.create<tensor::SplatOp>(loc, op.getQueryType(), scale);
+      scaled_query = rewriter.create<arith::MulFOp>(loc, query, scaleTensor);
+    } else {
+      dk = std::sqrt(dk) / M_LOG2E;
+      Value scale = rewriter.create<arith::ConstantOp>(
+          loc, targetType, rewriter.getFloatAttr(targetType, dk));
+      Value scaleTensor =
+          rewriter.create<tensor::SplatOp>(loc, op.getQueryType(), scale);
+      scaled_query = rewriter.create<arith::DivFOp>(loc, query, scaleTensor);
+    }
```

Comment:
We don't use arith ops directly on tensors. Can you use a linalg op to do this instead?

Reply:
Okay,
Do you think `tensor::SplatOp`` also be replaced with `linalg` ops?
`Value scaleTensor = rewriter.create<tensor::SplatOp>(loc, op.getQueryType(), scale);`

Reply:
```mlir
Value emptyTensor = rewriter.create<tensor::EmptyOp>(
      loc, op.getQueryType().getShape(), targetType);
      auto fillOp = rewriter.create<linalg::FillOp>(loc, scale, emptyTensor);
      auto scaleTensor = fillOp.getResult(0);
```